### PR TITLE
KAFKA-4501: Java 9 compilation and runtime fixes

### PR DESCRIPTION
**Pull request by ijuma**
Compilation error fixes:
- Avoid ambiguity error when appending to Properties in Scala
code (https://github.com/scala/bug/issues/10418)
- Use position() and limit() to fix ambiguity issue (
https://github.com/scala/bug/issues/10418#issuecomment-316364778)
- Disable findBugs if Java 9 is used (
https://github.com/findbugsproject/findbugs/issues/105)

Compilation warning fixes:
- Avoid deprecated Class.newInstance in Utils.newInstance
- Silence a few Java 9 deprecation warnings
- var -> val and unused fixes

Runtime error fixes:
- Introduce Base64 class that works in Java 7 and Java 9

Also:
- Set --release option if building with Java 9

Note that tests involving EasyMock (https://github.com/easymock/easymock/issues/193)
or PowerMock (https://github.com/powermock/powermock/issues/783)
will fail as neither supports Java 9 currently.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6636/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by ijuma** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6637/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6652/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6661/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by ijuma** [comment]
Review by @hachikuji.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6646/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6694/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6679/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by ijuma** [comment]
@hachikuji, https://builds.apache.org/job/kafka-trunk-jdk9/1/console shows that the build compiles with Java 9, but EasyMock/PowerMock tests fail.

File: core/src/main/scala/kafka/log/LogConfig.scala [hunk 1]
```diff
@@ -269,8 +270,8 @@ object LogConfig {
    */
   def fromProps(defaults: java.util.Map[_ <: Object, _ <: Object], overrides: Properties): LogConfig = {
     val props = new Properties()
-    props.putAll(defaults)
-    props.putAll(overrides)
+    defaults.asScala.foreach { case (k, v) => props.put(k, v) }
```
Comment (hachikuji) [inline core/src/main/scala/kafka/log/LogConfig.scala:273]: Curious why you didn't add another implicit `++=` for this?

File: clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslClientProvider.java [hunk 2]
```diff
@@ -25,10 +25,11 @@
 
     private static final long serialVersionUID = 1L;
 
+    @SuppressWarnings("deprecation")
     protected ScramSaslClientProvider() {
         super("SASL/SCRAM Client Provider", 1.0, "SASL/SCRAM Client Provider for Kafka");
         for (ScramMechanism mechanism : ScramMechanism.values())
-            super.put("SaslClientFactory." + mechanism.mechanismName(), ScramSaslClientFactory.class.getName());
+            put("SaslClientFactory." + mechanism.mechanismName(), ScramSaslClientFactory.class.getName());
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslClientProvider.java:32]: Why was this needed?

File: clients/src/main/java/org/apache/kafka/common/utils/Base64.java [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+                try {
+                    juUrlEncoderNoPassing = lookup.findVirtual(juEncoderClass, "withoutPadding",
+                            MethodType.methodType(juEncoderClass)).invoke(getUrlEncoder.invoke());
+                } catch (Throwable throwable) {
+                    // Invoked method doesn't throw checked exceptions, so safe to cast
+                    throw (RuntimeException) throwable;
+                }
+                URL_ENCODE_NO_PADDING = encode.bindTo(juUrlEncoderNoPassing);
+
+
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/utils/Base64.java]: nit: do we need the extra newlines?

**Review by hachikuji — COMMENTED** [review]
Seems straightforward. Just a couple questions.

File: core/src/main/scala/kafka/log/LogConfig.scala (hunk 1 above)
Comment (ijuma) [reply core/src/main/scala/kafka/log/LogConfig.scala:273]: The key type is Object instead of String, which is unsafe. We should probably change the type of the map received by LogConfig but I didn't want to tackle that here.

File: clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslClientProvider.java (hunk 2 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslClientProvider.java:32]: The super qualifier didn't achieve anything.

File: clients/src/main/java/org/apache/kafka/common/utils/Base64.java (hunk 3 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/utils/Base64.java]: Unintentional, will fix.

**Review by hachikuji — APPROVED** [review]
LGTM. Will let you merge when you fix the unneeded newlines.

**Comment by ijuma** [comment]
Fixed the empty lines, merging to trunk.
